src/Server.cpp

```diff
@@ -0,0 +1,211 @@
+        }
+
+        for (int n = 0; n < nfds; ++n) {
+            if (_events[n].data.fd == _serverSock)
+                acceptConnection();
+            else
+                handleConnections(n);
+        }
+    }
+}
+
+/**
+ * @brief Accept new connections
+ * 
+ * A new connection is accepted, its address saved in _cliendAddr and a 
+ * file descriptor for the new client socket is saved in clientSock. The 
+ * clientSock is added to the list of fds watched by epoll.
+ * 
+ * The while-loop allows for multiple connections to be accepted in one call
+ * of this function. It is broken when accept sets errno to EAGAIN or 
+ * EWOULDBLOCK (equivalent error codes indicating that a non-blocking operation 
+ * would normally block), meaning that no more connections are pending.
+ */
+void    Server::acceptConnection(){
+    while (1) {
+        socklen_t	addr_len = sizeof(_clientAddr);
+        int	clientSock = accept(
+            _serverSock, (struct sockaddr*)&_clientAddr, &addr_len);
+        if (clientSock < 0) {
+            if (errno == EAGAIN || errno == EWOULDBLOCK) {
+                break; // No more pending connections
+            } else {
+                std::cerr << "error: accept\n";
+                break;
+            }
+        }
+
+        // Set client socket to non-blocking
+        setNonblocking(clientSock);
+
+        // Add client socket to epoll instance
+        _ev.events = EPOLLIN | EPOLLET;
+        _ev.data.fd = clientSock;
+        if (epoll_ctl(_epfd, EPOLL_CTL_ADD, clientSock, &_ev) < 0) {
+            std::cerr << "error: epoll_ctl: clientSock\n";
+            close(clientSock);
+        }
+    }
+}
+
+/**
+ * @brief Handle data from a client connection.
+ * 
+ * This method is responsible for reading data from a client socket
+ * and processing it.
+ * 
+ * @param index The index of the event in the epoll events array.
+ * 
+ * @details The method performs the following steps:
+ * 
+ * 1. Retrieves the file descriptor of the client socket from the epoll events array.
+ * 2. Enters a loop to continuously read data from the client socket.
+ * 3. Reads data from the client socket using the read function into a buffer.
+ *    - If bytesRead is less than 0, it indicates that there is no more data to be read
+ *      or an error occurred with the read operation, and the loop breaks.
+ *    - If bytesRead is 0, it indicates that the connection has been closed by the client,
+ *      so the method closes the client socket and exits the loop.
+ *    - Otherwise, the method echoes the received data back to the client by writing it
+ *      to the client socket using the write function.
+ *  
+ */
+void    Server::handleConnections(int index){
+     // Handle client data
+    int clientSock = _events[index].data.fd;
+    while (1) {
+        char	buffer[BUFFER_SIZE];
+        int		bytesRead = read(clientSock, buffer, BUFFER_SIZE);
+        if (bytesRead < 0) {
+			// No more data to be read or error with read
+			break;
```

Comment:
This confuses "actual" read errors with non-blocking EAGAIN, in effect trying to do what we are not allowed.

Like I proposed above we shouldn't set the client sock to non-blocking since we can't check for errors anyway (we're not allowed to check errno so we can't differentiate between "good" EAGAIN and "bad" all others).

I think the way we should / have to handle it is 
- read the data into a buffer 
- check the buffer if it contains a full request. 
- if yes parse request
- if no we need to leave and come again - concatenate the buffer next time

Reply:
Implemented change to:

- blocking client sockets
- level-triggered events for client sockets in epoll

Results in:

- no while loop which would result in a blocked server
- read errors can be handled again
- repeated call of handleConnection() required to read complete request
- additional string member m_requestString to retain read data
- concatenation of buffer to m_requestString